include/updater/source.hpp

```diff
@@ -0,0 +1,96 @@
+#ifndef OSRM_UPDATER_SOURCE_HPP
+#define OSRM_UPDATER_SOURCE_HPP
+
+#include "util/typedefs.hpp"
+
+#include <boost/optional.hpp>
+
+#include <vector>
+
+namespace osrm
+{
+namespace updater
+{
+
+template <typename Key, typename Value> struct LookupTable
+{
+    boost::optional<Value> operator()(const Key &key) const
+    {
+        using Result = boost::optional<Value>;
+        const auto it = std::lower_bound(
+            lookup.begin(), lookup.end(), key, [](const auto &lhs, const auto &rhs) {
+                return rhs < lhs.first;
+            });
+        return it != std::end(lookup) && !(it->first < key) ? Result(it->second) : Result();
+    }
+
+    std::vector<std::pair<Key, Value>> lookup;
+};
+
+struct Segment final
+{
+    std::uint64_t from, to;
+    Segment() : from(0), to(0) {}
+    Segment(const std::uint64_t from, const std::uint64_t to) : from(from), to(to) {}
+    Segment(const OSMNodeID from, const OSMNodeID to)
+        : from(static_cast<std::uint64_t>(from)), to(static_cast<std::uint64_t>(to))
+    {
+    }
+
+    bool operator<(const Segment &rhs) const
+    {
+        return std::tie(from, to) < std::tie(rhs.from, rhs.to);
+    }
+    bool operator==(const Segment &rhs) const
+    {
+        return std::tie(from, to) == std::tie(rhs.from, rhs.to);
+    }
+};
+
+struct SpeedSource final
+{
+    SpeedSource() : speed(0), weight(std::numeric_limits<double>::quiet_NaN()) {}
```

Comment:
This is from the original code. `NaN` signifies "no value set" and means it will compute the weight from the duration value.